File: frontend/lib/models/video.dart
class Video {
  final int id;
  final int moduleId;
  final String title;
  final String? description;
  final String? youtubeVideoId; // Only available after access check
  final int duration;
  final int order;
  final bool isPreview;
  final bool? isLocked;
  final bool? isCompleted;
  final double? completionPercentage;
  final int? watchedSeconds;

  Video({
    required this.id,
    required this.moduleId,
    required this.title,
    this.description,
    this.youtubeVideoId,
    required this.duration,
    required this.order,
    required this.isPreview,
    this.isLocked,
    this.isCompleted,
    this.completionPercentage,
    this.watchedSeconds,
  });

  factory Video.fromJson(Map<String, dynamic> json) {
    return Video(
      id: json['id'],
      moduleId: json['module_id'],
      title: json['title'],
      description: json['description'],
      youtubeVideoId: json['youtube_video_id'],
      duration: json['duration'] ?? 0,
      order: json['order'] ?? 0,
      isPreview: json['is_preview'] == true || json['is_preview'] == 1,
      isLocked: json['is_locked'] != null
          ? (json['is_locked'] == true || json['is_locked'] == 1)
          : null,
      isCompleted: json['is_completed'] != null
          ? (json['is_completed'] == true || json['is_completed'] == 1)
          : null,
      completionPercentage: json['completion_percentage'] != null
          ? double.tryParse(json['completion_percentage'].toString())
          : null,
      watchedSeconds: json['watched_seconds'],
    );
  }

  /// Format duration as "12:30"
  String get formattedDuration {
    final minutes = duration ~/ 60;
    final seconds = duration % 60;
    return '${minutes}:${seconds.toString().padLeft(2, '0')}';
  }

  /// Video status icon
  String get statusEmoji {
    if (isCompleted == true) return '✅';
    if (isLocked == true) return '🔒';
    if (isPreview) return '▶️';
    return '🔓';
  }
}

File: backend/app/Models/Video.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Video extends Model
{
    use HasFactory;

    /**
     * The attributes that are mass assignable.
     */
    protected $fillable = [
        'module_id',
        'title',
        'description',
        'youtube_video_id',
        'duration',
        'order',
        'is_preview',
        'is_active',
    ];

    protected $casts = [
        'duration' => 'integer',
        'order' => 'integer',
        'is_preview' => 'boolean',
        'is_active' => 'boolean',
    ];

    /**
     * The attributes that should be hidden for serialization.
     * YouTube video ID is hidden by default — only returned after access check.
     */
    protected $hidden = [
        'youtube_video_id',
    ];

    /**
     * Get the module that owns the video.
     */
    public function module(): BelongsTo
    {
        return $this->belongsTo(Module::class);
    }

    /**
     * Get progress records for this video.
     */
    public function progress(): HasMany
    {
        return $this->hasMany(VideoProgress::class);
    }

    /**
     * Get the previous video in the same module (by order).
     */
    public function getPreviousVideo(): ?Video
    {
        return Video::where('module_id', $this->module_id)
            ->where('order', '<', $this->order)
            ->orderBy('order', 'desc')
            ->first();
    }

    /**
     * Get the previous video in the entire course sequence (across all modules).
     */
    public function getPreviousVideoInCourse(): ?Video
    {
        $courseId = $this->module->course_id;

        // Get all videos in this course ordered by module order and video order
        $videos = Video::whereHas('module', fn($q) => $q->where('course_id', $courseId))
            ->with('module')
            ->get()
            ->sortBy([
                fn($a, $b) => $a->module->order <=> $b->module->order,
                fn($a, $b) => $a->order <=> $b->order,
            ])
            ->values();

        $currentIndex = $videos->search(fn($v) => $v->id === $this->id);

        if ($currentIndex === false || $currentIndex === 0) {
            return null;
        }

        return $videos[$currentIndex - 1];
    }

    /**
     * Check if this is the first video in its module.
     */
    public function isFirstInModule(): bool
    {
        return !Video::where('module_id', $this->module_id)
            ->where('order', '<', $this->order)
            ->exists();
    }

    /**
     * Get the formatted duration (e.g., "12:30").
     */
    public function getFormattedDurationAttribute(): string
    {
        $minutes = floor($this->duration / 60);
        $seconds = $this->duration % 60;
        return sprintf('%d:%02d', $minutes, $seconds);
    }
}
